File: src/tools/exa_tool.py
"""Exa API wrapper for neural web search."""

import httpx
from typing import List, Dict, Any, Optional
from src.utils.config import Config

# Exa's /search endpoint validates `type` against this enum. Any other value
# yields HTTP 400. Callers (e.g. the planning agent) sometimes pass content-type
# values like "news" or "research paper" intending content filtering — those
# belong in `category`, not `type`. We sanitize at this boundary.
VALID_EXA_TYPES = frozenset({
    "neural", "keyword", "auto", "hybrid", "fast",
    "deep-reasoning", "deep-lite", "magic", "deep", "instant",
})


class ExaTool:
    """Wrapper for Exa API endpoints."""

    def __init__(self):
        """Initialize Exa API client."""
        self.api_key = Config.EXA_API_KEY
        self.base_url = Config.EXA_BASE_URL
        self.headers = {
            "x-api-key": self.api_key,
            "Content-Type": "application/json",
        }

    def search(
        self,
        query: str,
        num_results: int = 10,
        start_published_date: Optional[str] = None,
        end_published_date: Optional[str] = None,
        include_domains: Optional[List[str]] = None,
        exclude_domains: Optional[List[str]] = None,
        type: str = "auto",
        use_autoprompt: bool = True,
        text: bool = True,
        highlights: bool = True,
    ) -> Dict[str, Any]:
        """
        Perform neural search using Exa API.

        Args:
            query: Search query
            num_results: Number of results to return
            start_published_date: Filter results published after this date (ISO format)
            end_published_date: Filter results published before this date (ISO format)
            include_domains: List of domains to include
            exclude_domains: List of domains to exclude
            type: Search algorithm — one of VALID_EXA_TYPES (auto, neural, keyword, hybrid, fast, deep-reasoning, deep-lite, magic, deep, instant). Out-of-set values silently fall back to "auto".
            use_autoprompt: Let Exa optimize the query
            text: Include full text content
            highlights: Include relevant highlights

        Returns:
            Dictionary containing search results
        """
        url = f"{self.base_url}/search"

        # Sanitize `type`: Exa rejects unknown values with HTTP 400. Fall back
        # to "auto" so the planning agent's content-type strings ("news",
        # "research paper", "pdf") degrade gracefully instead of crashing.
        if type not in VALID_EXA_TYPES:
            type = "auto"

        payload = {
            "query": query,
            "numResults": num_results,
            "useAutoprompt": use_autoprompt,
            "type": type,
            "contents": {
                "text": text,
                "highlights": highlights,
            },
        }

        if start_published_date:
            payload["startPublishedDate"] = start_published_date
        if end_published_date:
            payload["endPublishedDate"] = end_published_date
        if include_domains:
            payload["includeDomains"] = include_domains
        if exclude_domains:
            payload["excludeDomains"] = exclude_domains

        try:
            with httpx.Client(timeout=30.0) as client:
                response = client.post(url, json=payload, headers=self.headers)
                response.raise_for_status()
                return response.json()
        except httpx.HTTPError as e:
            return {
                "error": str(e),
                "status": "failed",
                "results": [],
            }

    def find_similar(
        self,
        url: str,
        num_results: int = 5,
        exclude_source_domain: bool = True,
        text: bool = True,
        highlights: bool = True,
    ) -> Dict[str, Any]:
        """
        Find similar content to a given URL using Exa's neural similarity search.

        Args:
            url: URL to find similar content for
            num_results: Number of similar results to return
            exclude_source_domain: Exclude results from the same domain
            text: Include full text content
            highlights: Include relevant highlights

        Returns:
            Dictionary containing similar search results
        """
        api_url = f"{self.base_url}/findSimilar"

        payload = {
            "url": url,
            "numResults": num_results,
            "excludeSourceDomain": exclude_source_domain,
            "contents": {
                "text": text,
                "highlights": highlights,
            },
        }

        try:
            with httpx.Client(timeout=30.0) as client:
                response = client.post(api_url, json=payload, headers=self.headers)
                response.raise_for_status()
                return response.json()
        except httpx.HTTPError as e:
            return {
                "error": str(e),
                "status": "failed",
                "results": [],
            }

    def get_contents(
        self,
        ids: List[str],
        text: bool = True,
        highlights: bool = True,
    ) -> Dict[str, Any]:
        """
        Get full content for specific result IDs.

        Args:
            ids: List of Exa result IDs
            text: Include full text content
            highlights: Include relevant highlights

        Returns:
            Dictionary containing content for the specified IDs
        """
        url = f"{self.base_url}/contents"

        payload = {
            "ids": ids,
            "contents": {
                "text": text,
                "highlights": highlights,
            },
        }

        try:
            with httpx.Client(timeout=30.0) as client:
                response = client.post(url, json=payload, headers=self.headers)
                response.raise_for_status()
                return response.json()
        except httpx.HTTPError as e:
            return {
                "error": str(e),
                "status": "failed",
                "contents": [],
            }

    def format_results(self, results: Dict[str, Any]) -> List[Dict[str, Any]]:
        """
        Format Exa search results into a standardized structure.

        Args:
            results: Raw results from Exa API

        Returns:
            List of formatted result dictionaries
        """
        if "error" in results or "results" not in results:
            return []

        formatted = []
        for result in results.get("results", []):
            formatted_result = {
                "title": result.get("title", "No title"),
                "url": result.get("url", ""),
                "author": result.get("author"),
                "published_date": result.get("publishedDate"),
                "score": result.get("score", 0),
                "text": result.get("text", ""),
                "highlights": result.get("highlights", []),
                "summary": result.get("summary", ""),
            }
            formatted.append(formatted_result)

        return formatted

File: src/tools/exa_tool.ts
/** Exa API wrapper for neural web search. */

export interface ExaResult {
  id?: string;
  title?: string;
  url?: string;
  text?: string;
  publishedDate?: string;
  highlights?: string[];
  score?: number;
  author?: string;
  summary?: string;
}

export interface ExaResponse {
  results?: ExaResult[];
  error?: string;
  status?: string;
}

export interface ExaSearchOptions {
  numResults?: number;
  type?: "auto" | "keyword" | "neural";
  useAutoprompt?: boolean;
  startPublishedDate?: string;
  endPublishedDate?: string;
  includeDomains?: string[];
  excludeDomains?: string[];
  includeText?: boolean;
  includeHighlights?: boolean;
}

export interface ExaFindSimilarOptions {
  numResults?: number;
  excludeSourceDomain?: boolean;
  includeText?: boolean;
  includeHighlights?: boolean;
}

export interface ExaGetContentsOptions {
  includeText?: boolean;
  includeHighlights?: boolean;
}

export interface ExaToolOptions {
  apiKey: string;
  baseUrl?: string;
  fetchImpl?: typeof fetch;
}

export class ExaTool {
  private apiKey: string;
  private baseUrl: string;
  private fetchImpl: typeof fetch;

  constructor(opts: ExaToolOptions) {
    this.apiKey = opts.apiKey;
    this.baseUrl = opts.baseUrl ?? "https://api.exa.ai";
    this.fetchImpl = opts.fetchImpl ?? fetch;
  }

  async search(query: string, options: ExaSearchOptions = {}): Promise<ExaResponse> {
    const payload: Record<string, unknown> = {
      query,
      numResults: options.numResults ?? 10,
      type: options.type ?? "auto",
      useAutoprompt: options.useAutoprompt ?? true,
      contents: {
        text: options.includeText ?? true,
        highlights: options.includeHighlights ?? true,
      },
    };
    if (options.startPublishedDate) payload["startPublishedDate"] = options.startPublishedDate;
    if (options.endPublishedDate) payload["endPublishedDate"] = options.endPublishedDate;
    if (options.includeDomains?.length) payload["includeDomains"] = options.includeDomains;
    if (options.excludeDomains?.length) payload["excludeDomains"] = options.excludeDomains;
    return await this.post("/search", payload);
  }

  async findSimilar(url: string, options: ExaFindSimilarOptions = {}): Promise<ExaResponse> {
    return await this.post("/findSimilar", {
      url,
      numResults: options.numResults ?? 5,
      excludeSourceDomain: options.excludeSourceDomain ?? true,
      contents: {
        text: options.includeText ?? true,
        highlights: options.includeHighlights ?? true,
      },
    });
  }

  async getContents(
    ids: string[],
    options: ExaGetContentsOptions = {},
  ): Promise<ExaResponse> {
    return await this.post("/contents", {
      ids,
      contents: {
        text: options.includeText ?? true,
        highlights: options.includeHighlights ?? true,
      },
    });
  }

  /** Format results as a human-readable string. Returns "" for error responses. */
  formatResults(response: ExaResponse): string {
    if (response.error) return `Error: ${response.error}`;
    if (!response.results?.length) return "No results found.";
    const lines: string[] = [];
    for (const r of response.results) {
      lines.push(`Title: ${r.title ?? "(no title)"}`);
      if (r.url) lines.push(`URL: ${r.url}`);
      if (r.publishedDate) lines.push(`Published: ${r.publishedDate}`);
      if (r.text) lines.push(`Excerpt: ${r.text.slice(0, 500)}`);
      lines.push("");
    }
    return lines.join("\n");
  }

  private async post(path: string, body: unknown): Promise<ExaResponse> {
    try {
      const res = await this.fetchImpl(`${this.baseUrl}${path}`, {
        method: "POST",
        headers: {
          "x-api-key": this.apiKey,
          "Content-Type": "application/json",
        },
        body: JSON.stringify(body),
      });
      if (!res.ok) {
        return {
          error: `Exa API ${path} failed: ${res.status} ${res.statusText}`,
          status: "failed",
          results: [],
        };
      }
      return (await res.json()) as ExaResponse;
    } catch (e) {
      return {
        error: `Exa API ${path} failed: ${e instanceof Error ? e.message : String(e)}`,
        status: "failed",
        results: [],
      };
    }
  }
}
